### ISSUE TITLE
How to capture netmap bridge incoming packets

### ISSUE THREAD

**Issue by JamesZHuang** [luigirizzo/netmap#248]
Run example program bridge (on CentOS 6.5) e.g.
   `bridge -v netmap:eth0 netmap:eth0^`

could only see 'net->host' messages, they are outgoing bound packets.
There are no 'host->net' messages show up, where should be incoming bound packets.
However, the traffic goes through properly, i.e. run ping command for next hop router, and proved by tshark outputs.

How could we capture the incoming packets?
What we try to accomplish is, filter incoming packets from NIC, capture interested ones, manipulate them, then forward back to kernel stack (host). We care about performance and want to take advantage of zero copy,  is this correct direction to do it based on bridge example?

**Comment by vmaffione** [comment]
Yes, you can do it. This discussion already come out recenty (see recent issues on this repo), and people already did this, transforming the bridge into a packet filter for the network stack.

**Comment by JamesZHuang** [comment]
app bridge can only see one direction packets as `net->host`. 

tshark outputs is:

    ...
     70.420091 10.222.1.201 -> 10.222.1.47  ICMP 74 Echo (ping) request  id=0x0200, seq=48164/9404, ttl=128
     70.420143  10.222.1.47 -> 10.222.1.201 ICMP 74 Echo (ping) reply    id=0x0200, seq=48164/9404, ttl=64
     70.420453  10.222.1.47 -> 10.222.1.201 ICMP 74 Echo (ping) reply    id=0x0200, seq=48164/9404, ttl=64
     71.420053 10.222.1.201 -> 10.222.1.47  ICMP 74 Echo (ping) request  id=0x0200, seq=48420/9405, ttl=128
     71.420096  10.222.1.47 -> 10.222.1.201 ICMP 74 Echo (ping) reply    id=0x0200, seq=48420/9405, ttl=64
     71.420244  10.222.1.47 -> 10.222.1.201 ICMP 74 Echo (ping) reply    id=0x0200, seq=48420/9405, ttl=64
    ...
`10.222.1.47` is local ip, `10.222.1.201` is remote ip.
The duplicate outgoing reply packets are seen in bridge, while the incoming request packet is NOT captured in bridge.

I scrolled all existing issues in this repo, no luck to find answer. If I missed please point out. 
Have I missed something to capture the incoming packets?

**Comment by JamesZHuang** [comment]
is it possible related to too old CentOS 6.5 OS kernel 2.6.32-431.5.1.el6.x86_64?

**Comment by vmaffione** [comment]
Regarding the issue for which you don't see host->net, I answered here https://github.com/luigirizzo/netmap/issues/51.

This issue is related to what you are trying to do is here https://github.com/luigirizzo/netmap/issues/213

**Comment by JamesZHuang** [comment]
After I did #51 fix in `bridge`, it does change the printing to `host->net`, but no `net->host` prints.
I run #213 example single rx queue program with interface `eth0-0/R`, it does not report any incoming packets.

    nm_open [883] eth0-0/R tx 0 .. 0 1 rx 0 .. 0 1
    nm_open [886] TX0 0xe2fda400 h 0 c 1 t 1
    nm_open [886] TX1 0xe2fda400 h 0 c 1 t 1
    nm_open [890] RX0 0xe2ff3000 h 0 c 0 t 0
    nm_open [890] RX1 0xe2fda400 h 0 c 1 t 1
    [+] Receiving packets on interface eth0, RX ring 0
    [+] receiving 0 pps
    [+] receiving 0 pps
    [+] receiving 0 pps
    [+] receiving 0 pps

I run `test_nm eth0 read` , it does not receive any incoming packets as well.

When I run above tools, underneath, incoming packets do forward into host, as network traffic behaviours normal on the system, ping works, ssh works, udp works. But netmap can't capture the incoming packets. Why?

**Comment by vmaffione** [comment]
With the syntax `eth0-0/R` you are capturing only the packet received on the first rx ring of eth0. Maybe all the packets are passing through the second rx ring, which you don't touch.

What do you want to do, exactly?

**Comment by JamesZHuang** [comment]
It seems on my testbox, only `eth0-0/R` is allowed, `eth0-1/R` is invalid

    NIOCREGIF failed: Invalid argument eth0-1/R
    
I want to filter all incoming packets from nic (e.g. eth0),  directly forward un-interested packets back to host/kernel.  in the meantime capture interested packets for application processing then send back to host/kernel stack.

I can not see any incoming packets on polling `eth0 POLLIN` from netmap, I do see packets on polling `eth0^ POLLIN`.

**Comment by vmaffione** [comment]
Use pkt-gen as a test tool, don't use test_nm.
Why don't you use the simpler syntax eth0/R?

e.g.
````
# pkt-gen -i eth0/R -f rx
````
will read all the packets incoming on eth0.

you can use

````
# bridge eth0 eth0^
````
to intercept packets from eth0 and forward to the host stack. You can modify bridge to filter the unwanted packets.

**Comment by JamesZHuang** [comment]
Hi Vincenzo,

i tried everything i can (e.g. pkt-gen, bridge, etc), there are no incoming packets captured on eth0 on CentOS 6.5 with kernel 2.6.32-431.5.1.el6.x86_64.

I doubt it might be kernel version related and tried on CentOS 7.0 with kernel 3.10.0-123.el7.x86_64. Yes, it works, there are incoming packets captured on eth0. 

Unfortunately, our application was deployed on CentOS 6.5 and there is no plan to upgrade the OS. Did you experience similar scenario? could you please guide a direction how could I find the root cause and fix it?

**Comment by vmaffione** [comment]
Hi @JamesZHuang ,
  Sorry, I was not realizing it was a problem related to the kernel version.
Ok, so you are saying you're using the generic (aka emulated) driver, right? I imagine the NIC you are using does not have netmap native support (e.g. like Intel).

Does the transmission work?

````
# pkt-gen -i eth0 -f tx -R 100
````
?

Have you checked that eth0 is not attached to some kind of bridge (e.g. with brctl, openvswitch)?

**Comment by JamesZHuang** [comment]
Thank you understand my problem.
Yes, i'm using generic driver.
it seems pkt-gen transmission `-f tx` working fine.
the eth0 runs alone, does not attach to anything else.

but, it seems "`pkt-gen -i eth0 -f rx`" receives nothing, got repeating message

      receiver_body [1386] waiting for initial packets, poll returns 0 0

**Comment by vmaffione** [comment]
Ok, so clearly it is not working, since I guess you are sending packets to eth0 from an external machine. Can you paste the netmap_linux_config.h produced by ./configure on the problematic machine?

**Comment by JamesZHuang** [comment]
[LINUX]$ cat ./netmap_linux_config.h
        #ifndef NETMAP_LINUX_CONFIG_H
        #define NETMAP_LINUX_CONFIG_H

                #define NETMAP_LINUX_HAVE_E1000E_EXT_RXDESC

                #define NETMAP_LINUX_HAVE_SPLIT_PAGE

                #define NETMAP_LINUX_HAVE_SKB_CHECKSUM_START_OFFSET

                #define NETMAP_LINUX_HAVE_SKB_FRAG_ADDRESS

                #define NETMAP_LINUX_HAVE_SKB_FRAG_SIZE

                #define NETMAP_LINUX_SKB_ADD_RX_FRAG_6ARGS

                #define NETMAP_LINUX_HAVE_ALLOC_SKB_IP_ALIGN

                #define NETMAP_LINUX_HAVE_ADDR_RANDOM

                #define NETMAP_LINUX_HAVE_TX_SKB_SHARING

                #define NETMAP_LINUX_HAVE_LIVE_ADDR_CHANGE

                #define NETMAP_LINUX_HAVE_HRTIMER_MODE_REL

                #define NETMAP_LINUX_HAVE_USLEEP_RANGE

                #define NETMAP_LINUX_HAVE_NETDEV_TX_T

                #define NETMAP_LINUX_HAVE_NETDEV_OPS

                #define NETMAP_LINUX_HAVE_PHYS_ADDR_T

                #define NETMAP_LINUX_HAVE_HRTIMER_FORWARD_NOW

                #define NETMAP_LINUX_HAVE_QUEUE_MAPPING

                #define NETMAP_LINUX_HAVE_UINTPTR

                #define NETMAP_LINUX_HAVE_SKB_COPY_LINEAR

                #define NETMAP_LINUX_HAVE_INIT_NET

                #define NETMAP_LINUX_HAVE_UNLOCKED_IOCTL

                #define NETMAP_LINUX_PWAIT_KEY key

                #define NETMAP_LINUX_HAVE_GET_RINGPARAM

                #define NETMAP_LINUX_SELECT_QUEUE 2

                #define NETMAP_LINUX_TIMER_RTYPE enum hrtimer_restart

                #define NETMAP_LINUX_DRIVER_SUFFIX ""

        #endif

**Comment by vmaffione** [comment]
Ok, so the picture is clear.
As you can see, there is a very important item missing: `NETMAP_LINUX_HAVE_RX_REGISTER`. If you look at `LINUX/netmap_linux.c`, you will see that netmap intercepts the packets received from the driver using the netdev_rx_handler_register(). This functionality is obviously required for netmap generic reception to work. For such an old kernel, this functionality does not exist, so the nm_os_catch_rx() function does nothing.

In other words, reception in generic mode is not supported on 2.6.32. It may be that there is an equivalent functionality, maybe the function is just called differently. If you find this kind of equivalent functionality we can add support.

In the meanwhile I'll add a warning, so that the user is at least aware of the problem.

**Comment by JamesZHuang** [comment]
it's too bad. I tried to find an equivalent `netdev_rx_handler_register()`, but no luck, there is even no `rx_handler` field in struct `net_device`, it seems more than difficult to register a rx hook in kernel 2.6.32.
